### ISSUE TITLE
t.i.unix.Port fails on bound incoming unix sockets

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2002]
|[<img alt="ghazel's avatar" src="https://avatars.githubusercontent.com/u/31402?s=50" width="50" height="50">](https://github.com/ghazel)| @ghazel reported|
|-|-|
|Trac ID|trac#2002|
|Type|defect|
|Created|2006-08-17 23:50:22Z|

unix.Port specifically breaks connections from unix domain sockets that are bound.

t.i.u.Port:
```
    def _buildAddr(self, name):
        assert not name
        return None
```

<details><summary>Searchable metadata</summary>

```
trac-id__2002 2002
type__defect defect
reporter__ghazel ghazel
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1155858622000000 1155858622000000
changetime__1156757440000000 1156757440000000
version__None None
owner__ghazel ghazel
cc__exarkun cc__itamarst
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="ghazel's avatar" src="https://avatars.githubusercontent.com/u/31402?s=50" width="50" height="50">](https://github.com/ghazel)<a name="note_1"></a>|@ghazel commented|
|-|-|

unixpeer-2002 with unittest

**Comment by twisted-trac** [comment]
|[<img alt="ghazel's avatar" src="https://avatars.githubusercontent.com/u/31402?s=50" width="50" height="50">](https://github.com/ghazel)<a name="note_2"></a>|@ghazel commented|
|-|-|

After this, bindAddress support should be added to connectUNIX.

**Comment by twisted-trac** [comment]
|[<img alt="ghazel's avatar" src="https://avatars.githubusercontent.com/u/31402?s=50" width="50" height="50">](https://github.com/ghazel)<a name="note_3"></a>|@ghazel commented|
|-|-|

It looks like something is wrong on OS X with this branch..

Is there any reason twisted.internet.unix would not be used by OS X?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Dropping review keyword until things are sorted out on OS X

**Comment by twisted-trac** [comment]
|[<img alt="ghazel's avatar" src="https://avatars.githubusercontent.com/u/31402?s=50" width="50" height="50">](https://github.com/ghazel)<a name="note_5"></a>|@ghazel commented|
|-|-|

Works on OS X now.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @ghazel<br>@exarkun set status to `new`|
|-|-|

testPeerBind should have a docstring describing the feature it is testing.

I think self._addPorts(l) should be called regardless of whether the factory's Deferred fires successfully.  If it isn't, if the test fails, it will also leak a port object.  Likewise for the socket object the test creates.  ```done``` should probably be addBoth'd to the Deferred (and passthrough its argument, in case it is failure).

It's generally better if asserts are lexically contained by their test method, but since that doesn't seem to be the style in test_unix, the way you have written the test is okay.  Something to keep in mind in general, though.  

Go ahead and merge after adding a docstring and fixing the cleanup issues.

**Comment by twisted-trac** [comment]
|[<img alt="ghazel's avatar" src="https://avatars.githubusercontent.com/u/31402?s=50" width="50" height="50">](https://github.com/ghazel)<a name="note_9"></a>|@ghazel commented|
|-|-|

The other tests also use addCallback, I was mimicing those.

I added the cleanup, but the other tests should be reviewed if this is actually an issue.

**Comment by twisted-trac** [comment]
|<img alt="alus's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|alus set status to `closed`|
|-|-|

(In [18009]) unix.Port getPeer support

Author: ghazel

Reviewer: exarkun

Fixes [#2002](https://github.com/twisted/twisted/issues/2002)

unix domain sockets can now get the remote endpoint that the connecting side is bound to using the normal method:
transport.getPeer(). includes unittest
